google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -3343,6 +3578,27 @@ private void createSessions(final int sessionCount, boolean distributeOverChanne
     }
   }
 
+  class MultiplexedSessionConsumer implements SessionConsumer {
+
+    @Override
+    public void onSessionReady(SessionImpl sessionImpl) {
+      final MultiplexedSession session = new MultiplexedSession(sessionImpl);
+      synchronized (lock) {
+        multiplexedSession.set(session);
+        numMultiplexedSessionsBeingCreated = false;
+        while (!multiplexedSessionWaiters.isEmpty()) {
+          multiplexedSessionWaiters.poll().put(multiplexedSession.get());
+        }
+      }
+    }
+
+    @Override
+    public void onSessionCreateFailure(Throwable t, int createFailureForSessionCount) {
+      numMultiplexedSessionsBeingCreated = false;
+      handleMultiplexedSessionsFailure(newSpannerException(t));
```

Comment:
What is the state of the session pool if this happens for the initial `CreateSession` RPC?

The curent implementation seems to get into a non-valid state, as the maintainer will not try to re-create the session, but the error that we get here is only being propagated to the current waiters. That means that threads that try to get a multiplexed session in the future will get a ... `NullPointerException`? Or something else?

I think that the following would be simpler, both from a code perspective and usability perspective:
1. Keep the current multiplexed session in a `SettableFuture`. 
2. Let this method call `SettableFuture.setException(..)`.
3. Let all waiters just wait on that future (so simply by calling `SettableFuture.get()`). That will fail automatically when this method calls `SettableFuture.setException(..)`.
4. All threads that in the future try to get a multiplexed session also automatically get the exception that was set here when they try to call `SettableFuture.get()`.

Reply:
Done, simplified the follow by using `SettableFuture`.

Reply:
Done, simplified this flow by using `AtomicReference<SettableApiFuture<MultiplexedSessionFuture>>.`